### ISSUE TITLE
Ability to select secondary GPU

### ISSUE THREAD

**Issue by K900** [ValveSoftware/steam-for-linux#634]
Now that we can have `optirun %command%`, can we have a way to enable it for all games at once? Not really important (well, not until we get more games anyway), just throwing it out there.

**Comment by OctoNezd** [comment]
It issue live for more than 4 years, maybe just add in config feature to use custom launch command?

**Comment by RussianNeuroMancer** [comment]
@kisak-valve

Are you sure that request to implement native DRI_PRIME support (which is works for dGPU and eGPU) is the same as request to implement optirun support?

**Comment by kisak-valve** [comment]
It's effectively the same request to have a select secondary gpu option in Steam. If or when there is a functional difference between gpu vendors, then we can re-evaluate having separate feature requests.

**Comment by mercuriete** [comment]
Hi @kisak-valve

It would be nice to have the feature of selecting vulkan driver provider via VK_ICD_FILENAMES
I already launch steam with only radv json and the application is forced to use secondary and powerful graphic card.

Thank you in advance.

**Comment by nbmvoid** [comment]
I honestly don't know where to ask this but I'm dealing with the same problem on Steam (PRIME offloading)

sh ~/prime.sh %command% Works perfectly if a game binary was launched (CS:GO)

It doesn't work correctly if the game is launched via a wrapper (such as Left 4 Dead 2, Goldsrc games, etc)

My prime.sh cointains this
`__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia __VK_LAYER_NV_optimus=NVIDIA_only VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/nvidia_icd.json "$@"`

**Comment by nathan-weinberg** [comment]
Running Fedora 31, would also like to see this feature.

**Comment by mercuriete** [comment]
Just as adding more information,
Mesa added a vulkan selection layer.
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/1766

DRI_PRIME=1 should work as well for vulkan in the near future.

Waiting for mesa 20.1.0 :tada: 

PS: I've never tried before but somebody can try laptopmodetools to export some variables when laptop is plugged in.

**Comment by crawfxrd** [comment]
[switcheroo-control](https://gitlab.freedesktop.org/hadess/switcheroo-control) provides a DBus interface for enumerating available GPUs. It will provide the environment variables needed to offload to them, and supports the NVIDIA proprietary drivers.

```
gdbus introspect --system --dest net.hadess.SwitcherooControl --object-path /net/hadess/SwitcherooControl --only-properties
```

```
node /net/hadess/SwitcherooControl {
  interface net.hadess.SwitcherooControl {
    properties:
      readonly b HasDualGpu = true;
      readonly u NumGPUs = 2;
      readonly aa{sv} GPUs = [{'Name': <'NVIDIA Corporation TU106M [GeForce RTX 2070 Mobile]'>, 'Environment': <['__GLX_VENDOR_LIBRARY_NAME', 'nvidia', '__NV_PRIME_RENDER_OFFLOAD', '1']>, 'Default': <false>}, {'Name': <'Intel Corporation UHD Graphics 630 (Mobile)'>, 'Environment': <['DRI_PRIME', 'pci-0000_00_02_0']>, 'Default': <true>}];
  };
};
```

GNOME uses this to provide the "Launch using Dedicated Graphics Card" on laptops configured for PRIME render offload.

**Comment by kisak-valve** [comment]
Add PrefersNonDefaultGPU=true to .desktop file of Steam

Issue transferred from https://github.com/ValveSoftware/steam-for-linux/issues/7089.
@Eonfge posted on 2020-05-06T14:52:23:

#### Please describe your issue in as much detail as possible:
There is a new standard for aiding in multi-GPU setups. You can now set a flag called `PrefersNonDefaultGPU` which will tell the computer to use the alternative (generally speaking, dedicated) GPU. 

By setting this flag, Steam and it's applications will run with the more powerful GPU by default, preventing many annoyed customers as they are rushing B with 20 fps.

#### Docs
- https://www.gamingonlinux.com/articles/the-linux-desktop-entry-specification-gets-a-way-to-automatically-use-a-discrete-gpu-merged-into-gnome.16598
- https://www.hadess.net/2020/05/dual-gpu-support-launch-on-discrete-gpu.html
- https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#recognized-keys

**Comment by crawfxrd** [comment]
Huh, I figured since Steam itself was running the dGPU it would prevent the card from suspending. But that seems to not be the case if it's minimized to the tray (nothing to draw, after all).

```
$ cat /sys/bus/pci/devices/0000\:01\:00.*/power/runtime_status
suspended
suspended
suspended
suspended
```

Anyway, we did just get support merged into GNOME master: [GNOME/gnome-shell!1226](https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1226)

**Comment by smcv** [comment]
steam-launcher beta 1.0.0.70 adds `PrefersNonDefaultGPU=true`, as well as the older `X-KDE-RunOnDiscreteGpu=true`.

(If you get your Steam client through an unofficial distro-specific package, they will need to update to 1.0.0.70 before you see this change.)
